### PR TITLE
[SYCL] Correct CFE behavior with named lambdas in unnamed-lambda mode

### PR DESCRIPTION
**Pull request by erichkeane**
In unnamed-lambda mode, we currently treat ALL kernels as using the
unnamed-lambda feature to the language, which is wrong.  Named lambdas
should still work like they did in SYCL2017, and the unnamed lambda functionality
only applies when the user omits the name.

The library is responsible to replace a non-named kernel's name with the
type of the lambda.  However, it still allows the user to specify the
name if they'd like, and uses that as the name instead.

Because of this, we can identify the unnamed lambda by when the first
template argument and the functor are the same unqualified type, meaning
the library has done the above replacement.

There are a couple of features in this patch that all need to work together/
be merged together since they all break eachother.

First is marking of the kernel functors.  The current implementation ALWAYS 
marked the kernel functor as the name
of the kernel, but this isn't correct, it isn't the name of the kernel
unless the library makes it so.  This ends up causing user-named kernels
that have their names changed by the functor they pass in.

Second is the 'default' value of LangOpts.SYCLUnnamedLambdas.  When not
specified, we should be in unnamed-lambda mode if we are in SYCL2020.  This
patch also corrects that, PLUS adds an -fno flag for the same thing.  This also
required fixing up how -sycl-std is handled (since it isn't quite a lang-standard,
and can't be treated directly as a lang-arg, since others can depend on it).

Third is SemaSYCL Sema checking.  We need to still check bad kernel names when
in unnamed-lambda mode if the kernel name is user provided.

Forth is the Integration header.  There is an 'unnamed lambda' format for it and a
'normal' format for it.  We need to make sure we generate the correct one for each
type of functor.

### CODE REVIEW

**Comment by erichkeane** [comment]
@AlexeySachkov

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -1011,16 +1011,19 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
       .Default(InvokeUnknown);
 }
 
-static const CXXRecordDecl *getKernelObjectType(FunctionDecl *Caller) {
-  assert(Caller->getNumParams() > 0 && "Insufficient kernel parameters");
+// The SYCL kernel's 'object type' used for diagnostics and naming/mangling is
+// the first parameter to a sycl_kernel labeled function template. In SYCL1.2.1,
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:1015]:
```suggestion
// the first parameter to a function template using the sycl_kernel attribute. In SYCL1.2.1,
```

File: clang/lib/Sema/SemaSYCL.cpp [hunk 2]
```diff
@@ -3683,10 +3686,12 @@ void Sema::CheckSYCLKernelCall(FunctionDecl *KernelFunc, SourceRange CallLoc,
...
   if (!LangOpts.SYCLIsDevice)
     return;
-  const CXXRecordDecl *KernelObj = getKernelObjectType(KernelFunc);
+  QualType KernelObjType = GetSYCLKernelObjectType(KernelFunc);
   QualType KernelNameType =
       calculateKernelNameType(getASTContext(), KernelFunc);
-  if (!KernelObj) {
+  const CXXRecordDecl *KernelObj;
+  if (KernelObjType.isNull() ||
+      !(KernelObj = KernelObjType->getAsCXXRecordDecl())) {
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:3694]:
```suggestion
  const CXXRecordDecl *KernelObj = !KernelObjType.isNull() ? KernelObjType->getAsCXXRecordDecl()) : nullptr;  
  if (!KernelObj) {
```
Though, of more interest to me is: can that `QualType` ever actually *be* null? I would assume that if it's null, that means there was some sort of error we already reported, but then why are we getting to *here* and giving a second error?

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:3694]: I don't have a good idea of when it would happen actually... I couldn't come up with any examples, so i think it does make sense to just not have this condition and count on the `operator->` to cause the assert until we find it.  A non-asserts build will still emit the diagnostic though.

**Review by AaronBallman — DISMISSED** [review]
LGTM from an FE perspective.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 3]
```diff
@@ -3683,9 +3687,11 @@ void Sema::CheckSYCLKernelCall(FunctionDecl *KernelFunc, SourceRange CallLoc,
   // issues.
   if (!LangOpts.SYCLIsDevice)
     return;
-  const CXXRecordDecl *KernelObj = getKernelObjectType(KernelFunc);
+  const CXXRecordDecl *KernelObj =
+      GetSYCLKernelObjectType(KernelFunc)->getAsCXXRecordDecl();
   QualType KernelNameType =
       calculateKernelNameType(getASTContext(), KernelFunc);
+  ;
```
[the remark this answers is not in the document]
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:3694]: Woopsie!

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -1011,16 +1011,20 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
       .Default(InvokeUnknown);
 }
 
-static const CXXRecordDecl *getKernelObjectType(FunctionDecl *Caller) {
-  assert(Caller->getNumParams() > 0 && "Insufficient kernel parameters");
+// The SYCL kernel's 'object type' used for diagnostics and naming/mangling is
+// the first parameter to a function template using the sycl_kernel
+// attribute labeled function template. In SYCL1.2.1, this was passed by value,
```
Comment (premanandrao) [inline clang/lib/Sema/SemaSYCL.cpp:1016]:
```suggestion
// attribute. In SYCL1.2.1, this was passed by value,
```

**Review by premanandrao — COMMENTED** [review]
Just a nit.

**Comment by erichkeane** [comment]
Latest push pulls in https://github.com/intel/llvm/pull/3974 which between the two, should work if @AlexeySachkov  and my expectations are correct.

**Comment by erichkeane** [comment]
Hmm... that doesn't look right, going to do a force-push or two if necessary.

**Comment by erichkeane** [comment]
There, I think that is correct, @AlexeySachkov  this rebase seemed to skip the merges you did on the other branch, so if any of those are necessary let me know and I'll work on doing that too.

**Comment by rolandschulz** [comment]
Only reviewed changes to the SYCL headers. LGTM.

**Comment by erichkeane** [comment]
@AlexeySachkov  : looks like your assert hits on a lot of these?

**Comment by AlexeySachkov** [comment]
> @AlexeySachkov : looks like your assert hits on a lot of these?

@erichkeane, I think that https://github.com/intel/llvm/pull/3974/commits/6ebbe1cd0ee4801429c8d32431a083802ea6568a is missing

**Comment by erichkeane** [comment]
> 
> 
> > @AlexeySachkov : looks like your assert hits on a lot of these?
> 
> @erichkeane, I think that [6ebbe1c](https://github.com/intel/llvm/commit/6ebbe1cd0ee4801429c8d32431a083802ea6568a) is missing

Oh, thanks, I think you're right!  I have the 'test', but not the contents of the patch.  I must have messed up the conflict resolution there!

**Comment by gmlueck** [comment]
I know this isn't directly related to this change, but I saw that you added `-fsycl-unnamed-lambda` to a bunch of the tests.  Why do we require users to specify a compiler option in order to enable the unnamed lambda support?  Unnamed lambdas are an official part of SYCL 2020, so they are not an extension.  Why wouldn't we want them enabled by default?

**Comment by erichkeane** [comment]
> 
> 
> I know this isn't directly related to this change, but I saw that you added `-fsycl-unnamed-lambda` to a bunch of the tests. Why do we require users to specify a compiler option in order to enable the unnamed lambda support? Unnamed lambdas are an official part of SYCL 2020, so they are not an extension. Why wouldn't we want them enabled by default?

https://github.com/intel/llvm/issues/4009

**Comment by erichkeane** [comment]
> 
> 
> I know this isn't directly related to this change, but I saw that you added `-fsycl-unnamed-lambda` to a bunch of the tests. Why do we require users to specify a compiler option in order to enable the unnamed lambda support? Unnamed lambdas are an official part of SYCL 2020, so they are not an extension. Why wouldn't we want them enabled by default?

Also note these are being added to the cc1 invocation, which is after the driver has run, which typically sets the 'language defaults'.  Depending on how #4009 gets fixed we may or may not need these.

**Comment by erichkeane** [comment]
So I worked most of the day on trying to make the unnamed-lambdas' default-enabled in SYCL2020, however have found that the feature is either not well implemented, or I don't understand what is going on.

Turning on the unnamed lambdas seems to force us to ignore the user-provided name in just about every way, and I think our implementation design REQUIRES that thanks to how the integration-header stuff works, and how some of the type checking works.

**Comment by erichkeane** [comment]
Note I just updated the commit message, this patch is getting quite sizable unfortunately.  We need to fix how we manage unnamed kernels throughout the compiler.  We apparently assumed that once the flag is on that ALL kernels are unnamed, but this isn't the case.

**Comment by erichkeane** [comment]
Note all the int-header stuff is going to fail, and anything that uses the int-footer.  I still have to correctly implement the integration-header for named lambdas in unnamed-configurations.

**Comment by erichkeane** [comment]
@v-klochkov : I believe there is some issue with the reduction.hpp implementation (that git-blame is targetting you:)) where it isn't correctly maintaining the 'unnamed-kernel-ness' of a call.  For a call to remain an unnamed kernel (and thus opt into the 'old' int-header stuff, and to opt-out of the error-checking) the 1st template argument must be the same as the type of the functor.

Until this patch, we haven't enforced this, so we've gotten away with this issue in the get_reduction_aux_kernel_name_t stuff.  Could you take a look here?

Thanks!

**Comment by erichkeane** [comment]
@v-klochkov : Pushed b1df7ea to test my hypothesis.

File: clang/include/clang/Driver/Options.td [hunk 5]
```diff
@@ -2605,9 +2606,15 @@ def fsycl_link_EQ : Joined<["-"], "fsycl-link=">,
...
-  MarshallingInfoFlag<LangOpts<"SYCLUnnamedLambda">>;
+defm sycl_unnamed_lambda
+    : BoolFOption<"sycl-unnamed-lambda", LangOpts<"SYCLUnnamedLambda">,
+                  Default<!strconcat(
+                      sycl_ver.KeyPath,
+                      " >= clang::LangOptions::SYCLMajorVersion::SYCL_2020")>,
+                  PosFlag<SetTrue, [CC1Option, CoreOption],
+                          "Allow unnamed SYCL lambda kernels">,
+                  NegFlag<SetFalse, [CC1Option, CoreOption],
+                          "Disallow unnamed SYCL lambda kernels">>;
```
Comment (mdtoguchi) [inline clang/include/clang/Driver/Options.td:2617]:
```suggestion
                  BothFlags<[CC1Option, CoreOption],
                          " unnamed SYCL lambda kernels">,
                  PosFlag<SetTrue, [], "Allow">,
                  NegFlag<SetFalse, [], "Disallow">>;
```

**Comment by v-klochkov** [comment]
> @v-klochkov : Pushed [b1df7ea](https://github.com/intel/llvm/commit/b1df7eaa8ad11ff874e8b3db6f415a8b71a40555) to test my hypothesis.

@erichkeane - thank you for fixing that. The fix though needs some additional changes as it does not changes the class __sycl_reduction_main_kernel, which now has <typename, bool, bool, typename> and the proposed fix added an additional typename into that template-arg-list making it <typename, typename, bool, bool, typename>.

**Comment by erichkeane** [comment]
> ome additional changes as it does not changes the class __sycl_reduction_main_kernel, which now has <typename, bool, bool, typename> and the proposed fix added an additional typename into that template-arg-list making it <typename,

Ah, thanks!  I didn't really get it building last night apparently, but I'll work to correct it today.

**Comment by erichkeane** [comment]
Whew!  This finally got through testing :)  Please review it if you can!

File: clang/lib/Sema/SemaSYCL.cpp [hunk 6]
```diff
@@ -3208,6 +3212,18 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
...
 
+// Kernels are only the unnamed-lambda feature if the feature is enabled, AND
+// the first template argument has been corrected by the library to match the
+// functor type.
+static bool IsSYCLUnnamedKernel(Sema &SemaRef, const FunctionDecl *FD) {
+  if (!SemaRef.getLangOpts().SYCLUnnamedLambda)
+    return false;
+  QualType FunctorTy = GetSYCLKernelObjectType(FD);
+  QualType TmplArgTy =
+      calculateKernelNameType(SemaRef.Context, FD).getUnqualifiedType();
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:3223]: It's a little bit odd that one of these functions returns an unqualified type while the other one returns a possibly qualified type. Should `calculateKernelNameType()` always return an unqualified type instead? Or are there times when that qualification is important for the kernel name?

File: clang/lib/Sema/SemaSYCL.cpp [hunk 7]
```diff
@@ -4799,9 +4812,7 @@ void SYCLIntegrationHeader::setCallsThisGroup(bool B) {
   K->FreeFunctionCalls.CallsThisGroup = B;
 }
 
-SYCLIntegrationHeader::SYCLIntegrationHeader(bool _UnnamedLambdaSupport,
-                                             Sema &_S)
-    : UnnamedLambdaSupport(_UnnamedLambdaSupport), S(_S) {}
+SYCLIntegrationHeader::SYCLIntegrationHeader(Sema &_S) : S(_S) {}
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:4815]:
```suggestion
SYCLIntegrationHeader::SYCLIntegrationHeader(Sema &S) : S(S) {}
```
Let's not use a reserved identifier.

File: clang/test/CodeGenSYCL/kernel_functor.cpp [hunk 8]
```diff
@@ -1,5 +1,7 @@
-// RUN: %clang_cc1 -fsycl-is-device -fsycl-int-header=%t.h %s -o %t.spv
-// RUN: FileCheck %s --input-file=%t.h
+// RUN: %clang_cc1 -triple x86_64-linux-gnu -fsycl-is-device -fno-sycl-unnamed-lambda -fsycl-int-header=%t.h %s -o %t.spv
+// RUN: FileCheck %s --input-file=%t.h --check-prefixes=NUL,CHECK
+// RUN: %clang_cc1 -triple x86_64-linux-gnu -fsycl-is-device -fsycl-int-header=%t.h %s -o %t.spv
```
Comment (AaronBallman) [inline clang/test/CodeGenSYCL/kernel_functor.cpp:3]: Is `%t.spv` intended to be unused output?

File: clang/lib/Sema/SemaSYCL.cpp (hunk 6 above)
Comment (AaronBallman) [reply clang/lib/Sema/SemaSYCL.cpp:3223]: Also, can you double-check that we have tests for qualified kernel names and that we get the proper instantiations out of them?

File: clang/include/clang/Driver/Options.td [hunk 9]
```diff
@@ -2605,9 +2606,14 @@ def fsycl_link_EQ : Joined<["-"], "fsycl-link=">,
...
   HelpText<"Generate partially linked device object to be used with the host link">;
-def fsycl_unnamed_lambda : Flag<["-"], "fsycl-unnamed-lambda">,
-  Flags<[CC1Option, CoreOption]>, HelpText<"Allow unnamed SYCL lambda kernels">,
-  MarshallingInfoFlag<LangOpts<"SYCLUnnamedLambda">>;
+defm sycl_unnamed_lambda
+    : BoolFOption<
+          "sycl-unnamed-lambda", LangOpts<"SYCLUnnamedLambda">,
+          Default<!strconcat(
+              sycl_ver.KeyPath,
+              " >= clang::LangOptions::SYCLMajorVersion::SYCL_2020")>,
```
[the remark this answers is not in the document]
Comment (erichkeane) [reply clang/include/clang/Driver/Options.td:2614]: I don't know actually.  I suspect it WOULD work?  I figured though that sycl_ver is going to be useful to us in the future, so I thought it was a good thing to have anyway.

File: clang/include/clang/Driver/Options.td [hunk 10]
```diff
@@ -2605,9 +2606,14 @@ def fsycl_link_EQ : Joined<["-"], "fsycl-link=">,
...
+          PosFlag<SetTrue, [], "Allow">, NegFlag<SetFalse, [], "Disallow">,
+          BothFlags<[CC1Option, CoreOption], " unnamed SYCL lambda kernels">>;
```
[the remark this answers is not in the document]
Comment (erichkeane) [reply clang/include/clang/Driver/Options.td:2616]: I think it is?  This makes the documentation for the two flags to be:

`-fsycl-unnamed-kernels` == Allow unnamed SYCL lambda kernels
`-fno-sycl-unnamed-kernels` == Disallow unnamed SYCL lambda kernels

File: clang/lib/Sema/SemaSYCL.cpp (hunk 6 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:3223]: I think actually the unqualification here is wrong, we only want it in unqualified mode since we want to differentiate between:

`kernel<Foo>(...);`
and 
`kernel<const Foo>(...);`

So I'll remove the unqualification here.  As far as tests, I don't see any CFE tests, but I'm sure there has got to be one in at least one of the test suites SOMEWHERE.

File: clang/test/CodeGenSYCL/kernel_functor.cpp (hunk 8 above)
Comment (erichkeane) [reply clang/test/CodeGenSYCL/kernel_functor.cpp:3]: Yes, we only care about the int-header, but have to specify a -o anyway.  Presumably these could all just do a -o -, but I think that messes up the llvm-lit output.

File: sycl/test/functor/functor_inheritance.cpp [hunk 11]
```diff
@@ -1,4 +1,4 @@
-// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -o %t.out %s
```
Comment (v-klochkov) [inline sycl/test/functor/functor_inheritance.cpp:1]: This test does not need -fsycl-unnamed-lambda to pass. @erichkeane , why did you change this one?

Comment (v-klochkov) [reply sycl/test/functor/functor_inheritance.cpp:1]: Same question to kernel_functor.cpp, which probably also does not need -fsycl-unnamged-lambda option as well.
(but I did not try compiling that test with your changes).

File: clang/include/clang/Driver/Options.td (hunk 9 above)
[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/include/clang/Driver/Options.td:2614]: Hmmm, I can live with the variable, but I'm sometimes skeptical of "we may need this in the future" variables. I'd say keep it, though; I suspect you're right about using it more.

File: clang/include/clang/Driver/Options.td (hunk 10 above)
[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/include/clang/Driver/Options.td:2616]: Fantastic, thank you for double-checking for me!

File: clang/lib/Sema/SemaSYCL.cpp (hunk 6 above)
Comment (AaronBallman) [reply clang/lib/Sema/SemaSYCL.cpp]: We've thought that in the past :-D Can you make sure we have some test coverage for the case you mention? That seems like something we'd want to explicitly document as not naming the same kernel.

File: clang/test/CodeGenSYCL/kernel_functor.cpp (hunk 8 above)
Comment (AaronBallman) [reply clang/test/CodeGenSYCL/kernel_functor.cpp:3]: I think we usually do `-o /dev/null` so that we don't pay the cost of writing out to a temp file (and deleting it) that we don't care about.

File: sycl/test/functor/functor_inheritance.cpp (hunk 11 above)
Comment (erichkeane) [reply sycl/test/functor/functor_inheritance.cpp:1]: Because it used to :)  This patch unfortunately has grown a whole bunch as a result of seeing all the issues, so the flag wasn't 'default' when I first wrote this patch.  I'll audit/remove uses of it.

File: clang/test/CodeGenSYCL/kernel_functor.cpp (hunk 8 above)
Comment (erichkeane) [reply clang/test/CodeGenSYCL/kernel_functor.cpp:3]: Does that work on windows?!

**Review by v-klochkov — APPROVED** [review]

File: clang/include/clang/Driver/Options.td [hunk 12]
```diff
@@ -4647,7 +4653,7 @@ def fdebug_measure_parse_tree : Flag<["-"], "fdebug-measure-parse-tree">, Group<
   HelpText<"Measure the parse tree">;
 def fdebug_pre_fir_tree : Flag<["-"], "fdebug-pre-fir-tree">, Group<Action_Group>,
   HelpText<"Dump the pre-FIR tree">;
-def fdebug_module_writer : Flag<["-"],"fdebug-module-writer">, 
+def fdebug_module_writer : Flag<["-"],"fdebug-module-writer">,
```
Comment (AaronBallman) [inline clang/include/clang/Driver/Options.td:4656]: More unintended whitespace -- your editor loves to eat those! :-)

**Comment by AaronBallman** [comment]
The precommit test failure *smells* unrelated, but it'd be nice to rerun the tests to make sure it goes back to green.

File: clang/include/clang/Driver/Options.td (hunk 12 above)
Comment (erichkeane) [inline clang/include/clang/Driver/Options.td:4656]:
```suggestion
def fdebug_module_writer : Flag<["-"],"fdebug-module-writer">, 
```

**Review by v-klochkov — APPROVED** [review]

**Review by mdtoguchi — APPROVED** [review]

**Review by AaronBallman — APPROVED** [review]
LGTM!
